server/src/main/java/com/hedera/block/server/producer/ProducerBlockItemObserver.java

```diff
@@ -108,28 +115,59 @@ public void onNext(@NonNull final PublishStreamRequest publishStreamRequest) {
             publishStreamResponseObserver.onNext(errorResponse);
             LOGGER.log(System.Logger.Level.ERROR, "Exception thrown publishing BlockItem: ", io);
-
             LOGGER.log(System.Logger.Level.ERROR, "Shutting down the web server");
             serviceStatus.stopWebServer();
+        } catch (ParseException e) {
+            @NonNull final var errorResponse = buildErrorStreamResponse();
+            publishStreamResponseObserver.onNext(errorResponse);
+            LOGGER.log(
+                    System.Logger.Level.ERROR,
+                    "Error parsing inbound block item from a producer: "
+                            + publishStreamRequest.getBlockItem(),
+                    e);
+            serviceStatus.stopWebServer();
         }
     }
 
     @NonNull
-    private PublishStreamResponse buildSuccessStreamResponse(@NonNull final BlockItem blockItem)
-            throws IOException, NoSuchAlgorithmException {
-        @NonNull final ItemAcknowledgement itemAck = itemAckBuilder.buildAck(blockItem);
-        return PublishStreamResponse.newBuilder().setAcknowledgement(itemAck).build();
+    private com.hedera.hapi.block.protoc.PublishStreamResponse buildSuccessStreamResponse(
+            @NonNull final BlockItem blockItem) throws IOException, NoSuchAlgorithmException {
+
+        @NonNull final Acknowledgement ack = buildAck(blockItem);
+
+        return toProtocPublishStreamResponse(
+                PublishStreamResponse.newBuilder().acknowledgement(ack).build());
     }
 
     @NonNull
-    private static PublishStreamResponse buildErrorStreamResponse() {
+    private static com.hedera.hapi.block.protoc.PublishStreamResponse buildErrorStreamResponse() {
         // TODO: Replace this with a real error enum.
         @NonNull
         final EndOfStream endOfStream =
                 EndOfStream.newBuilder()
-                        .setStatus(PublishStreamResponseCode.STREAM_ITEMS_UNKNOWN)
+                        .status(PublishStreamResponseCode.STREAM_ITEMS_UNKNOWN)
                         .build();
-        return PublishStreamResponse.newBuilder().setStatus(endOfStream).build();
+        return toProtocPublishStreamResponse(
+                PublishStreamResponse.newBuilder().status(endOfStream).build());
+    }
+
+    /**
+     * Protected method meant for testing. Builds an Acknowledgement for the block item.
+     *
+     * @param blockItem the block item to build the Acknowledgement for
+     * @return the Acknowledgement for the block item
+     * @throws NoSuchAlgorithmException if the hash algorithm is not supported
+     */
+    @NonNull
+    protected Acknowledgement buildAck(@NonNull final BlockItem blockItem)
+            throws NoSuchAlgorithmException {
+        ItemAcknowledgement itemAck =
+                ItemAcknowledgement.newBuilder()
+                        // TODO: Replace this with a real hash generator
+                        .itemHash(Bytes.wrap(getFakeHash(blockItem)))
```

Comment:
You won't want to hash here, it should be passed in.  That's because we'll be hashing _a lot_ and need to avoid repeating the work.  Also we may change the way we calculate hashes over time.
At some point we may even want to look into using OpenCL for hashing, because GPUs and NPUs are vastly faster at hashing than CPUs, but only if we _batch_ the hashing into a large group, so we _might_ do the entire block worth of hashes all in one batch to offload the effort and make use of hardware resources that are otherwise idle.

Reply:
This is good context when we circle back for this work.